libbeat/keystore/file_keystore.go

```diff
@@ -0,0 +1,385 @@
+func (k *FileKeystore) Save() error {
+	k.Lock()
+	defer k.Unlock()
+
+	if k.dirty == false {
+		return nil
+	}
+
+	temporaryPath := fmt.Sprintf("%s.tmp", k.Path)
+	defer func() {
+		if _, err := os.Stat(temporaryPath); os.IsExist(err) {
+			os.Remove(temporaryPath)
+		}
+	}()
+
+	if _, err := os.Stat(temporaryPath); os.IsExist(err) {
+		return fmt.Errorf("temporary keystore exist at '%s', delete it and try again", temporaryPath)
+	}
+
+	w := bytes.NewBuffer([]byte{})
+	jsonEncoder := json.NewEncoder(w)
+	err := jsonEncoder.Encode(k.secrets)
+
+	if err != nil {
+		return fmt.Errorf("cannot serialize the keystore before saving it to disk: %v", err)
+	}
+
+	encrypted, err := k.encrypt(w)
+
+	if err != nil {
+		return fmt.Errorf("cannot encrypt the keystore: %v", err)
+	}
+
+	f, err := os.OpenFile(temporaryPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, filePermission)
+
+	if err != nil {
+		return fmt.Errorf("cannot open file to save the keystore to '%s', error: %s", k.Path, err)
+	}
+
+	defer f.Close()
+
+	f.Write(version)
+
+	base64Encoder := base64.NewEncoder(base64.StdEncoding, f)
+	io.Copy(base64Encoder, encrypted)
+	base64Encoder.Close()
+
+	err = os.Rename(temporaryPath, k.Path)
+
+	if err != nil {
+		return fmt.Errorf("cannot replace the existing keystore, with the new keystore file at '%s', error: %s", k.Path, err)
+	}
+
+	k.dirty = false
+	return nil
+}
+
+// List return the availables keys
+func (k *FileKeystore) List() ([]string, error) {
+	k.RLock()
+	defer k.RUnlock()
+
+	keys := make([]string, 0, len(k.secrets))
+
+	for key := range k.secrets {
+		keys = append(keys, key)
+	}
+
+	return keys, nil
+}
+
+// GetConfig returns common.Config representation of the key / secret pair to be merged with other
+// loaded configuration.
+func (k *FileKeystore) GetConfig() (*common.Config, error) {
+	k.RLock()
+	defer k.RUnlock()
+
+	configHash := make(map[string]interface{})
+
+	for key, secret := range k.secrets {
+		configHash[key] = string(secret.Value)
+	}
+
+	return common.NewConfigFrom(configHash)
+}
+
+// Create create an empty keystore, if the store already exist we will return an error.
+func (k *FileKeystore) Create(override bool) error {
+	k.Lock()
+
+	_, err := os.Stat(k.Path)
+
+	if os.IsNotExist(err) && !override {
+		return ErrAlreadyExists
+	}
+
+	k.secrets = make(map[string]serializableSecureString)
+	k.dirty = true
+	k.Unlock()
+	return k.Save()
+}
+
+// Exists return if the keystore is physically persisted on disk
+func (k *FileKeystore) Exists() bool {
+	k.Lock()
+	defer k.Unlock()
+
+	if _, err := os.Stat(k.Path); err == nil {
+		return true
+	}
+	return false
+}
+
+func (k *FileKeystore) load() error {
+	k.Lock()
+	defer k.Unlock()
+
+	// If the file doesn't exist we ignore the error and go ahead,
+	// we assume an empty keystore any other error will be reported upstream
+	if _, err := os.Stat(k.Path); os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	err := k.checkPermissions(k.Path)
```

Comment:
I am a bit on the fence for this, since the goal is to protect the keystore, but sincce the `isStrictPerms()` is already used in the config context which can also have secrets..